### DISCUSSION TITLE
Use Web Workers for hash decoding?

### DISCUSSION THREAD

**Discussion by felixranesberger** [johannschopplich/unlazy#23] [Ideas]
Decoding the image placeholder hashes is currently running on the main thread. 
Should we move the decoding part into a dedicated web worker, so that we have less stress on the main thread?

I could provide a pull request for this feature 👍

**Comment by johannschopplich** [comment]
Great idea, why not! I think we still should put it into a feature flag/option to conditionally enable or disable it. 

Sure, go for it. PR happily welcome. 🌻

**Reply by felixranesberger** [reply]
I began working on the pull request. 
The implementation should hopefully be pretty simple. 

The only issue is that the `createPngDataUri` functions, which are also exposed by the library to "external users", will become async, because we need to wait for the worker to process the message.

Using a feature flag to enable the worker will also cause the function to be async, because we probably need to import the necessary code asynchronously, so we don't have the hashing algorithms twice (in the worker and in the exposed  `createPngDataUri` function).

I've researched the support for web workers a bit and it seems, that it _should_ work in all cases. 
* https://caniuse.com/?search=web%20workers
* https://medium.com/@yinte1222/nuxt3-how-to-use-web-workers-in-a-nuxt-app-d03c1c8f2952
* https://stackblitz.com/github/vercel/next.js/tree/canary/examples/with-web-worker?file=README.md

I would suggest using the we workers as a default and releasing a new major version, when the changes are merged, since some users may be using the `createPngDataUri` functions and may need to adjust their implementation.

**Reply by felixranesberger** [reply]
Current implementation is here: https://github.com/johannschopplich/unlazy/pull/25/files
